File: platform/Avalonia/SweetEditor/EditorSettings.cs
namespace SweetEditor {
	public sealed class EditorSettings {
		private readonly SweetEditorControl editor;

		private float editorTextSize;
		private string fontFamily;
		private float scale = 1.0f;
		private FoldArrowMode foldArrowMode = FoldArrowMode.ALWAYS;
		private WrapMode wrapMode = WrapMode.NONE;
		private bool compositionEnabled;
		private float lineSpacingAdd;
		private float lineSpacingMult = 1.0f;
		private float contentStartPadding;
		private bool showSplitLine = true;
		private bool gutterSticky = true;
		private bool gutterVisible = true;
		private CurrentLineRenderMode currentLineRenderMode = CurrentLineRenderMode.BACKGROUND;
		private AutoIndentMode autoIndentMode = AutoIndentMode.KEEP_INDENT;
		private bool backspaceUnindent;
		private bool readOnly;
		private int maxGutterIcons;
		private long decorationScrollRefreshMinIntervalMs = 16L;
		private float decorationOverscanViewportMultiplier = 1.5f;

		internal EditorSettings(SweetEditorControl editor) {
			this.editor = editor;
			editorTextSize = editor.RendererInternal.EditorTextSize;
			fontFamily = editor.RendererInternal.FontFamily;
			compositionEnabled = editor.EditorCoreInternal.IsCompositionEnabled();
			backspaceUnindent = editor.EditorCoreInternal.IsBackspaceUnindent();
			readOnly = editor.EditorCoreInternal.IsReadOnly();
		}

		public void SetEditorTextSize(float size) {
			editorTextSize = size > 0 ? size : editorTextSize;
			editor.RendererInternal.SetEditorTextSize(editorTextSize);
			editor.EditorCoreInternal.OnFontMetricsChanged();
			editor.Flush();
		}

		public float GetEditorTextSize() => editorTextSize;

		public void SetFontFamily(string family) {
			if (string.IsNullOrWhiteSpace(family)) {
				return;
			}
			fontFamily = family.Trim();
			editor.RendererInternal.SetFontFamily(fontFamily);
			editor.EditorCoreInternal.OnFontMetricsChanged();
			editor.Flush();
		}

		public string GetFontFamily() => fontFamily;

		public void SetTypeface(string typeface) => SetFontFamily(typeface);

		public string GetTypeface() => GetFontFamily();

		public void SetScale(float scale) {
			this.scale = scale > 0 ? scale : 1.0f;
			editor.EditorCoreInternal.SetScale(this.scale);
			editor.RendererInternal.SetScale(this.scale);
			editor.EditorCoreInternal.OnFontMetricsChanged();
			editor.Flush();
		}

		public float GetScale() => scale;

		public void SetFoldArrowMode(FoldArrowMode mode) {
			foldArrowMode = NormalizeFoldArrowMode(mode);
			editor.EditorCoreInternal.SetFoldArrowMode((int)foldArrowMode);
			editor.Flush();
		}

		public FoldArrowMode GetFoldArrowMode() => foldArrowMode;

		public void SetWrapMode(WrapMode mode) {
			wrapMode = NormalizeWrapMode(mode);
			editor.EditorCoreInternal.SetWrapMode((int)wrapMode);
			editor.Flush();
		}

		public WrapMode GetWrapMode() => wrapMode;

		public void SetCompositionEnabled(bool enabled) {
			if (compositionEnabled == enabled) {
				return;
			}

			if (!enabled && editor.EditorCoreInternal.IsComposing()) {
				editor.EditorCoreInternal.CompositionCancel();
			}

			compositionEnabled = enabled;
			editor.EditorCoreInternal.SetCompositionEnabled(enabled);
			editor.Flush();
		}

		public bool IsCompositionEnabled() => compositionEnabled;

		public void SetLineSpacing(float add, float mult) {
			lineSpacingAdd = add;
			lineSpacingMult = mult <= 0 ? 1.0f : mult;
			editor.EditorCoreInternal.SetLineSpacing(lineSpacingAdd, lineSpacingMult);
			editor.Flush();
		}

		public float GetLineSpacingAdd() => lineSpacingAdd;

		public float GetLineSpacingMult() => lineSpacingMult;

		public void SetContentStartPadding(float padding) {
			contentStartPadding = padding < 0 ? 0 : padding;
			editor.EditorCoreInternal.SetContentStartPadding(contentStartPadding);
			editor.Flush();
		}

		public float GetContentStartPadding() => contentStartPadding;

		public void SetShowSplitLine(bool show) {
			showSplitLine = show;
			editor.EditorCoreInternal.SetShowSplitLine(show);
			editor.Flush();
		}

		public bool IsShowSplitLine() => showSplitLine;

		public void SetGutterSticky(bool sticky) {
			gutterSticky = sticky;
			editor.EditorCoreInternal.SetGutterSticky(sticky);
			editor.Flush();
		}

		public bool IsGutterSticky() => gutterSticky;

		public void SetGutterVisible(bool visible) {
			gutterVisible = visible;
			editor.EditorCoreInternal.SetGutterVisible(visible);
			editor.Flush();
		}

		public bool IsGutterVisible() => gutterVisible;

		public void SetCurrentLineRenderMode(CurrentLineRenderMode mode) {
			currentLineRenderMode = NormalizeCurrentLineRenderMode(mode);
			editor.EditorCoreInternal.SetCurrentLineRenderMode(currentLineRenderMode);
			editor.Flush();
		}

		public CurrentLineRenderMode GetCurrentLineRenderMode() => currentLineRenderMode;

		public void SetAutoIndentMode(AutoIndentMode mode) {
			autoIndentMode = NormalizeAutoIndentMode(mode);
			editor.EditorCoreInternal.SetAutoIndentMode((int)autoIndentMode);
			editor.Flush();
		}

		public AutoIndentMode GetAutoIndentMode() => autoIndentMode;

		public void SetBackspaceUnindent(bool enabled) {
			backspaceUnindent = enabled;
			editor.EditorCoreInternal.SetBackspaceUnindent(enabled);
		}

		public bool IsBackspaceUnindent() => backspaceUnindent;

		public void SetReadOnly(bool readOnly) {
			this.readOnly = readOnly;
			editor.EditorCoreInternal.SetReadOnly(readOnly);
			editor.Flush();
		}

		public bool IsReadOnly() => readOnly;

		public void SetMaxGutterIcons(int count) {
			maxGutterIcons = count < 0 ? 0 : count;
			editor.EditorCoreInternal.SetMaxGutterIcons(maxGutterIcons);
			editor.Flush();
		}

		public int GetMaxGutterIcons() => maxGutterIcons;

		public void SetDecorationScrollRefreshMinIntervalMs(long ms) {
			decorationScrollRefreshMinIntervalMs = ms < 0 ? 0L : ms;
			editor.RequestDecorationRefresh();
		}

		public long GetDecorationScrollRefreshMinIntervalMs() => decorationScrollRefreshMinIntervalMs;

		public void SetDecorationOverscanViewportMultiplier(float multiplier) {
			decorationOverscanViewportMultiplier = multiplier < 0 ? 0 : multiplier;
			editor.RequestDecorationRefresh();
		}

		public float GetDecorationOverscanViewportMultiplier() => decorationOverscanViewportMultiplier;

		private static FoldArrowMode NormalizeFoldArrowMode(FoldArrowMode mode) {
			return mode is FoldArrowMode.AUTO or FoldArrowMode.ALWAYS or FoldArrowMode.HIDDEN
				? mode
				: FoldArrowMode.ALWAYS;
		}

		private static WrapMode NormalizeWrapMode(WrapMode mode) {
			return mode is WrapMode.NONE or WrapMode.CHAR_BREAK or WrapMode.WORD_BREAK
				? mode
				: WrapMode.NONE;
		}

		private static CurrentLineRenderMode NormalizeCurrentLineRenderMode(CurrentLineRenderMode mode) {
			return mode is CurrentLineRenderMode.BACKGROUND or CurrentLineRenderMode.BORDER or CurrentLineRenderMode.NONE
				? mode
				: CurrentLineRenderMode.BACKGROUND;
		}

		private static AutoIndentMode NormalizeAutoIndentMode(AutoIndentMode mode) {
			return mode is AutoIndentMode.NONE or AutoIndentMode.KEEP_INDENT
				? mode
				: AutoIndentMode.KEEP_INDENT;
		}
	}
}

File: platform/Flutter/sweeteditor/lib/editor_settings.dart
import 'editor_core.dart' as core;

abstract class EditorSettingsHost {
  void applyTypography({
    required double textSize,
    required String fontFamily,
    required double scale,
  });

  void applyFoldArrowMode(core.FoldArrowMode mode);
  void applyWrapMode(core.WrapMode mode);
  void applyLineSpacing(double add, double mult);
  void applyContentStartPadding(double padding);
  void applyShowSplitLine(bool show);
  void applyGutterSticky(bool sticky);
  void applyGutterVisible(bool visible);
  void applyCurrentLineRenderMode(core.CurrentLineRenderMode mode);
  void applyAutoIndentMode(core.AutoIndentMode mode);
  void applyReadOnly(bool readOnly);
  void applyCompositionEnabled(bool enabled);
  void applyMaxGutterIcons(int count);
  void requestDecorationRefresh();
  void flushEditor();
}

/// Settings wrapper for the editor.
class EditorSettings {
  EditorSettings();

  double _textSize = 14;
  String _fontFamily = 'monospace';
  double _scale = 1.0;
  core.FoldArrowMode _foldArrowMode = core.FoldArrowMode.always;
  core.WrapMode _wrapMode = core.WrapMode.none;
  double _lineSpacingAdd = 0;
  double _lineSpacingMult = 1.0;
  double _contentStartPadding = 0;
  bool _showSplitLine = true;
  bool _gutterSticky = true;
  bool _gutterVisible = true;
  core.CurrentLineRenderMode _currentLineRenderMode =
      core.CurrentLineRenderMode.background;
  core.AutoIndentMode _autoIndentMode = core.AutoIndentMode.none;
  bool _readOnly = false;
  bool _compositionEnabled = false;
  int _maxGutterIcons = 0;
  int _decorationScrollRefreshMinIntervalMs = 16;
  double _decorationOverscanViewportMultiplier = 1.5;
  bool _textSizeCustomized = false;
  bool _fontFamilyCustomized = false;
  EditorSettingsHost? _host;

  void seedDefaults({required double textSize, required String fontFamily}) {
    if (!_textSizeCustomized) {
      _textSize = textSize;
    }
    if (!_fontFamilyCustomized) {
      _fontFamily = fontFamily;
    }
  }

  void bind(EditorSettingsHost host) {
    _host = host;
    _applyAll(host);
  }

  void unbind(EditorSettingsHost host) {
    if (identical(_host, host)) {
      _host = null;
    }
  }

  void setEditorTextSize(double size) {
    _textSize = size;
    _textSizeCustomized = true;
    _host?.applyTypography(
      textSize: _textSize,
      fontFamily: _fontFamily,
      scale: _scale,
    );
    _host?.flushEditor();
  }

  double getEditorTextSize() => _textSize;

  void setFontFamily(String fontFamily) {
    _fontFamily = fontFamily;
    _fontFamilyCustomized = true;
    _host?.applyTypography(
      textSize: _textSize,
      fontFamily: _fontFamily,
      scale: _scale,
    );
    _host?.flushEditor();
  }

  String getFontFamily() => _fontFamily;

  void setScale(double scale) {
    _scale = scale;
    _host?.applyTypography(
      textSize: _textSize,
      fontFamily: _fontFamily,
      scale: _scale,
    );
    _host?.flushEditor();
  }

  double getScale() => _scale;

  void setFoldArrowMode(core.FoldArrowMode mode) {
    _foldArrowMode = mode;
    _host?.applyFoldArrowMode(mode);
    _host?.flushEditor();
  }

  core.FoldArrowMode getFoldArrowMode() => _foldArrowMode;

  void setWrapMode(core.WrapMode mode) {
    _wrapMode = mode;
    _host?.applyWrapMode(mode);
    _host?.flushEditor();
  }

  core.WrapMode getWrapMode() => _wrapMode;

  void setLineSpacing(double add, double mult) {
    _lineSpacingAdd = add;
    _lineSpacingMult = mult;
    _host?.applyLineSpacing(add, mult);
    _host?.flushEditor();
  }

  double getLineSpacingAdd() => _lineSpacingAdd;
  double getLineSpacingMult() => _lineSpacingMult;

  void setContentStartPadding(double padding) {
    _contentStartPadding = padding.clamp(0, double.infinity);
    _host?.applyContentStartPadding(_contentStartPadding);
    _host?.flushEditor();
  }

  double getContentStartPadding() => _contentStartPadding;

  void setShowSplitLine(bool show) {
    _showSplitLine = show;
    _host?.applyShowSplitLine(show);
    _host?.flushEditor();
  }

  bool isShowSplitLine() => _showSplitLine;

  void setGutterSticky(bool sticky) {
    _gutterSticky = sticky;
    _host?.applyGutterSticky(sticky);
    _host?.flushEditor();
  }

  bool isGutterSticky() => _gutterSticky;

  void setGutterVisible(bool visible) {
    _gutterVisible = visible;
    _host?.applyGutterVisible(visible);
    _host?.flushEditor();
  }

  bool isGutterVisible() => _gutterVisible;

  void setCurrentLineRenderMode(core.CurrentLineRenderMode mode) {
    _currentLineRenderMode = mode;
    _host?.applyCurrentLineRenderMode(mode);
    _host?.flushEditor();
  }

  core.CurrentLineRenderMode getCurrentLineRenderMode() =>
      _currentLineRenderMode;

  void setAutoIndentMode(core.AutoIndentMode mode) {
    _autoIndentMode = mode;
    _host?.applyAutoIndentMode(mode);
  }

  core.AutoIndentMode getAutoIndentMode() => _autoIndentMode;

  void setReadOnly(bool readOnly) {
    _readOnly = readOnly;
    _host?.applyReadOnly(readOnly);
  }

  bool isReadOnly() => _readOnly;

  void setCompositionEnabled(bool enabled) {
    _compositionEnabled = enabled;
    _host?.applyCompositionEnabled(enabled);
    _host?.flushEditor();
  }

  bool isCompositionEnabled() => _compositionEnabled;

  void setMaxGutterIcons(int count) {
    _maxGutterIcons = count;
    _host?.applyMaxGutterIcons(count);
  }

  int getMaxGutterIcons() => _maxGutterIcons;

  void setDecorationScrollRefreshMinIntervalMs(int intervalMs) {
    _decorationScrollRefreshMinIntervalMs = intervalMs.clamp(0, 1 << 30);
    _host?.requestDecorationRefresh();
  }

  int getDecorationScrollRefreshMinIntervalMs() =>
      _decorationScrollRefreshMinIntervalMs;

  void setDecorationOverscanViewportMultiplier(double multiplier) {
    _decorationOverscanViewportMultiplier = multiplier.clamp(
      0,
      double.infinity,
    );
    _host?.requestDecorationRefresh();
  }

  double getDecorationOverscanViewportMultiplier() =>
      _decorationOverscanViewportMultiplier;

  void _applyAll(EditorSettingsHost host) {
    host.applyTypography(
      textSize: _textSize,
      fontFamily: _fontFamily,
      scale: _scale,
    );
    host.applyFoldArrowMode(_foldArrowMode);
    host.applyWrapMode(_wrapMode);
    host.applyLineSpacing(_lineSpacingAdd, _lineSpacingMult);
    host.applyContentStartPadding(_contentStartPadding);
    host.applyShowSplitLine(_showSplitLine);
    host.applyGutterSticky(_gutterSticky);
    host.applyGutterVisible(_gutterVisible);
    host.applyCurrentLineRenderMode(_currentLineRenderMode);
    host.applyAutoIndentMode(_autoIndentMode);
    host.applyReadOnly(_readOnly);
    host.applyCompositionEnabled(_compositionEnabled);
    host.applyMaxGutterIcons(_maxGutterIcons);
    host.requestDecorationRefresh();
    host.flushEditor();
  }
}
